src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala

```diff
@@ -34,49 +35,88 @@ class CassandraJournal extends AsyncWriteJournal with CassandraRecovery with Cas
   val preparedConfirmMessage = session.prepare(confirmMessage)
   val preparedDeletePermanent = session.prepare(deleteMessage)
-  val preparedSelectHeader = session.prepare(selectHeader).setConsistencyLevel(readConsistency)
   val preparedSelectMessages = session.prepare(selectMessages).setConsistencyLevel(readConsistency)
+  val preparedCheckInUse = session.prepare(selectInUse).setConsistencyLevel(readConsistency)
+  val preparedWriteInUse = session.prepare(writeInUse)
+  val preparedSelectHighestPartition = session.prepare(selectHighestSequence).setConsistencyLevel(readConsistency)
 
   def asyncWriteMessages(messages: Seq[AtomicWrite]): Future[Seq[Try[Unit]]] = {
-    val groupedStatements = messages.map(statementGroup)
-    val batchStatements = groupedStatements.map({
-      case Success(atomicWrite) =>
-        executeBatch(batch => atomicWrite.foreach(batch.add)).map(_ => Success(()))
-      case Failure(e) =>
-        Future.successful(Failure[Unit](e))
+    // we need to preserve the order / size of this sequence even though we don't map
+    // AtomicWrites 1:1 with a C* insert
+    val serialized = messages.map(aw => Try { SerializedAtomicWrite(
+        aw.payload.head.persistenceId,
+        aw.payload.map(pr => Serialized(pr.sequenceNr, persistentToByteBuffer(pr))))
     })
+    val result = serialized.map(a => a.map(_ => ()))
 
-    Future.sequence(batchStatements)
-  }
+    val byPersistenceId = serialized.collect({ case Success(caw) => caw }).groupBy(_.persistenceId).values
+    val boundStatements = byPersistenceId.map(statementGroup)
 
-  private def statementGroup(atomicWrite: AtomicWrite): Try[Seq[BoundStatement]] = Try {
-    atomicWrite.payload.flatMap { m =>
-      val pnr = partitionNr(m.sequenceNr)
-      val write = preparedWriteMessage.bind(m.persistenceId, pnr: JLong, m.sequenceNr: JLong, persistentToByteBuffer(m))
-      if (partitionNew(m.sequenceNr)) Seq(preparedWriteHeader.bind(m.persistenceId, pnr: JLong), write) else Seq(write)
+    val batchStatements = boundStatements.map({ unit =>
+        executeBatch(batch => unit.foreach(batch.add))
+    })
+    val promise: Promise[Seq[Try[Unit]]] = Promise[Seq[Try[Unit]]]()
+
+    Future.sequence(batchStatements).onComplete {
+      case Success(_) => promise.complete(Success(result))
+      case Failure(e) => promise.failure(e)
     }
+
+    promise.future
   }
 
-  private def asyncDeleteMessages(messageIds: Seq[MessageId]): Future[Unit] = executeBatch { batch =>
-    messageIds.foreach { mid =>
-      val stmt =
-        preparedDeletePermanent.bind(mid.persistenceId, partitionNr(mid.sequenceNr): JLong, mid.sequenceNr: JLong)
-      batch.add(stmt)
+  private def statementGroup(atomicWrites: Seq[SerializedAtomicWrite]): Seq[BoundStatement] = {
+    val maxPnr = partitionNr(atomicWrites.last.payload.last.sequenceNr)
+    val firstSeq: JLong = atomicWrites.head.payload.head.sequenceNr
+    val minPnr: JLong = partitionNr(firstSeq)
```

Comment:
To be consistent with other variable declarations, `JLong` type should not be used here (for both `firstSeq` and `minPnr`) but rather when calling `preparedWriteInUse.bind` below ...